src/Http/Middleware/CsrfProtectionMiddleware.php

```diff
@@ -0,0 +1,184 @@
+     * @param \Psr\Http\Message\ResponseInterface $response The response.
+     * @param callable $next Callback to invoke the next middleware.
+     * @return \Psr\Http\Message\ResponseInterface A response
+     */
+    public function __invoke(ServerRequestInterface &$request, ResponseInterface $response, $next)
+    {
+        $cookies = $request->getCookieParams();
+        $cookieData = null;
+        if (isset($cookies[$this->_config['cookieName']])) {
+            $cookieData = isset($cookies[$this->_config['cookieName']]);
+        }
+
+        if (!empty($cookieData)) {
+            $params = $request->getAttribute('params');
+            $params['_csrfToken'] = $cookieData;
+            $request = $request->withAttribute('params', $params);
+        }
+
+        $method = $request->getMethod();
+        if ($method === 'requested') {
+            return $next($request, $response);
+        }
+
+        if ($method === 'GET' && $cookieData === null) {
+            $this->_setCookie($request, $response);
+
+            return $next($request, $response);
+        }
+
+        $this->_validateAndUnsetTokenField($request);
+
+        return $next($request, $response);
+    }
+
+    /**
+     * Checks if the request is POST, PUT, DELETE or PATCH and validates the CSRF token
+     *
+     * @param \Cake\Http\ServerRequest $request The request object.
+     * @return void
+     */
+    protected function _validateAndUnsetTokenField(ServerRequest &$request)
+    {
+        if (in_array($request->getMethod(), ['PUT', 'POST', 'DELETE', 'PATCH']) || $request->getData()) {
+            $this->_validateToken($request);
+            $body = $request->getParsedBody();
+            if (is_array($body)) {
+                unset($body[$this->_config['field']]);
+                $request = $request->withParsedBody($body);
+            }
+        }
+    }
+
+    /**
+     * Set the cookie in the response.
+     *
+     * Also sets the request->params['_csrfToken'] so the newly minted
+     * token is available in the request data.
+     *
+     * @param \Cake\Http\ServerRequest $request The request object.
+     * @param \Cake\Http\Response $response The response object.
+     * @return void
+     */
+    protected function _setCookie(ServerRequest &$request, Response &$response)
+    {
+        $expiry = new Time($this->_config['expiry']);
+        $value = hash('sha512', Security::randomBytes(16), false);
+
+        $params = $request->getAttribute('params');
+        $params['_csrfToken'] = $value;
+        $request = $request->withAttribute('params', $params);
+
+        $response = $response->withCookie($this->_config['cookieName'], [
+            'value' => $value,
+            'expire' => $expiry->format('U'),
+            'path' => $request->getAttribute('webroot'),
+            'secure' => $this->_config['secure'],
+            'httpOnly' => $this->_config['httpOnly'],
+        ]);
+    }
```

Comment:
No, because `_setCookie(ServerRequest &$request, Response &$response)`. I've tried to return an array with both objects before but this leads into failing tests because the reference gets lost somewhere. Feel free to try it and remove the `&` and return this array:

```php
        return [
            'request' => $request,
            'response' => $response
        ];
```